services/spar/test/Test/Spar/Sem/IdPSpec.hs

```diff
@@ -9,64 +10,237 @@ import Imports
+propsForInterpreter interpreter lower = do
+  describe interpreter $ do
+    prop "deleteConfig/deleteConfig" $ prop_deleteDelete lower
+    prop "deleteConfig/getConfig" $ prop_deleteGet lower
+    prop "getConfig/StoreConfig" $ prop_getStore lower
+    prop "getConfig/getConfig" $ prop_getGet lower
+    prop "setReplacedBy/cleatReplacedBy" $ prop_setClear lower
+    prop "setReplacedBy/getReplacedBy" $ prop_setGet lower
+    prop "setReplacedBy/setReplacedBy" $ prop_setSet lower
     prop "storeConfig/getConfig" $ prop_storeGet lower
     prop "storeConfig/getIdByIssuerWithoutTeam" $ prop_storeGetByIssuer lower
+    prop "storeConfig/storeConfig (different keys)" $ prop_storeStore lower
+    prop "storeConfig/storeConfig (same keys)" $ prop_storeStore lower
 
 spec :: Spec
 spec = modifyMaxSuccess (const 1000) $ do
-  propsForInterpreter testInterpreter
+  propsForInterpreter "idPToMem" $ pure . run . idPToMem
+
+getReplacedBy :: Member E.IdP r => SAML.IdPId -> Sem r (Maybe (Maybe SAML.IdPId))
+getReplacedBy idpid = fmap (view $ SAML.idpExtraInfo . IP.wiReplacedBy) <$> E.getConfig idpid
+
+-- | All the constraints we need to generalize properties in this module.
+-- A regular type synonym doesn't work due to dreaded impredicative
+-- polymorphism.
+class
+  (Member E.IdP r, forall z. Show z => Show (f z), forall z. Eq z => Eq (f z)) =>
+  PropConstraints r f
+
+instance
+  (Member E.IdP r, forall z. Show z => Show (f z), forall z. Eq z => Eq (f z)) =>
+  PropConstraints r f
+
+prop_storeStore ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeStore =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    s' <- arbitrary
+    pure
+      ( do
+          E.storeConfig $ s & SAML.idpId .~ s' ^. SAML.idpId
+          E.storeConfig s',
+        do
+          E.storeConfig s'
+      )
+
+prop_storeStoreInterleave ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeStoreInterleave =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    s' <- arbitrary
+    !_ <-
+      if s ^. SAML.idpId == s' ^. SAML.idpId
+        then discard
+        else pure ()
+    pure
+      ( do
+          E.storeConfig s
+          E.storeConfig s',
+        do
+          E.storeConfig s'
+          E.storeConfig s
+      )
+
+prop_storeGet ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeGet =
+  prepropLaw @'[E.IdP] $
+    do
+      s <- arbitrary
+      pure
+        ( do
+            E.storeConfig s
+            E.getConfig $ s ^. idpId,
+          do
+            E.storeConfig s
+            pure (Just s)
+        )
+
+prop_deleteGet ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_deleteGet =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    pure
+      ( do
+          E.deleteConfig s
+          E.getConfig $ s ^. SAML.idpId,
+        do
+          E.deleteConfig s
+          pure Nothing
```

Comment:
I think this test may be pointless if you haven't been careful about the Arbitrary instance.  Does it return the same UUID twice in a few calls with non-negligible probability?

Reply:
You're right --- it almost certainly doesn't. What would you say are the odds of the generators respecting the size parameter?

Reply:
I have no idea.  :)  experiment?  I'm not sure that's the right approach, though.  We want the keys to come from a small set so that collisions are common, but not anything else.  If we didn't use quickcheck but [registry](https://hackage.haskell.org/package/registry), we could "easily" replace the underlying UUID and Issuer generators, but with quickcheck I'm not sure what to do.

(Note to self: the Issuer generator may have not enough entropy, instead of too much.  With some of the generators I've been quite lazy.)